test/test_robin_dict.jl

```diff
@@ -0,0 +1,413 @@
+        f(2)
+    end == 4
+
+    @test get!(d, 42) do  # d is not updated
+        f(200)
+    end == 4
+
+    @test get(d, 13) do   # d is not updated
+        f(4)
+    end == 16
+
+    @test d == RobinDict(8=>19, 19=>2, 42=>4)
+end
+
+@testset "push!" begin
+    d = RobinDict()
+    @test push!(d, 'a' => 1) === d
+    @test d['a'] == 1
+    @test push!(d, 'b' => 2, 'c' => 3) === d
+    @test d['b'] == 2
+    @test d['c'] == 3
+    @test push!(d, 'd' => 4, 'e' => 5, 'f' => 6) === d
+    @test d['d'] == 4
+    @test d['e'] == 5
+    @test d['f'] == 6
+    @test length(d) == 6
+end
+
+@testset "pop!" begin
+    d = RobinDict(1=>2, 3=>4)
+    @test pop!(d, 1) == 2
+    @test_throws KeyError pop!(d, 1)
+    @test pop!(d, 1, 0) == 0
+    @test pop!(d) == (3=>4)
+    @test_throws ArgumentError pop!(d)
+end
+
+@testset "keys as a set" begin
+    d = RobinDict(1=>2, 3=>4)
+    @test keys(d) isa AbstractSet
+    @test empty(keys(d)) isa AbstractSet
+    let i = keys(d) ∩ Set([1,2])
+        @test i isa AbstractSet
+        @test i == Set([1])
+    end
+    @test Set(string(k) for k in keys(d)) == Set(["1","3"])
+end
+
+@testset "find" begin
+    @test findall(isequal(1), RobinDict(:a=>1, :b=>2)) == [:a]
+    @test sort(findall(isequal(1), RobinDict(:a=>1, :b=>1))) == [:a, :b]
+    @test isempty(findall(isequal(1), RobinDict()))
+    @test isempty(findall(isequal(1), RobinDict(:a=>2, :b=>3)))
+
+    @test findfirst(isequal(1), RobinDict(:a=>1, :b=>2)) == :a
+    @test findfirst(isequal(1), RobinDict(:a=>1, :b=>1, :c=>3)) in (:a, :b)
+    @test findfirst(isequal(1), RobinDict()) === nothing
+    @test findfirst(isequal(1), RobinDict(:a=>2, :b=>3)) === nothing
+end
+
+@testset "haskey" begin
+    h = RobinDict(1=>2, 2=>3)
+    @test haskey(h, 1) == true
+    @test haskey(h, 2) == true
+    @test haskey(h, 3) == false
+    @test_throws MethodError haskey(h, "1")
+end
+
+@testset "getkey" begin
+    h = RobinDict(1=>2, 3 => 6, 5=>10)
+    @test getkey(h, 1, 7) == 1
+    @test getkey(h, 4, 6) == 6
+    @test_throws MethodError getkey(h, "1", 8)
+end
+
+@testset "empty" begin
+    h = RobinDict()
+    for i=1:1000
+        h[i] = i+1
+    end
+    length0 = length(h.hashes)
+    empty!(h)
+    @test h.count == 0
+    @test h.maxprobe == 0
+    @test h.idxfloor == 0
+    @test h.totalcost == 0
+    @test length(h.hashes) == length(h.keys) == length(h.vals) == length0
+end
+
+@testset "ArgumentError" begin
+    @test_throws ArgumentError RobinDict(0)
+    @test_throws ArgumentError RobinDict([1])
+    @test_throws ArgumentError RobinDict([(1,2),0])
+end
+
+@testset "empty tuple" begin
+    h = RobinDict(())
+    @test length(h) == 0
+end
+
+@testset "get_idxfloor" begin
+    h = RobinDict()
+    @test get_idxfloor(h) == 0
+    
+    h["a"] = 1
+    h[2] = "b"
+    @test h.idxfloor == get_idxfloor(h)
+    pop!(h)
+    @test h.idxfloor == get_idxfloor(h)
+    pop!(h)
+    @test h.idxfloor == get_idxfloor(h) == 0 
+end
+
+@testset "invariants" begin
```

Comment:
Very good. So, what I had in mind was something like
```
function check_invariants(h::RobinDict)
cnt = 0
sz = length(h.keys)
for i=1:length(h.keys)
isslotfilled(h, i) || continue
@assert hash_key(h.keys[i]) == h.hashes[i]
@assert (h.hashes[i] & 0x80000000) != 0
cnt += 1
des_ind = desired_index(h.hashes[i], sz)
dist = calculate_distance(h, i)
@assert dist <= h.maxprobe
if dist == des_ind - i
for j = des_ind : (i-1)
@assert  calculate_distance(h, j)>= dist
else
@assert  dist == des_ind - i + sz
for j=des_ind:sz
@assert  calculate_distance(h, j)>= dist
end
for j = 1 : (i-1)
@assert  calculate_distance(h, j)>= dist
end
end
end
@assert cnt == length(h)
end
```
This would be to make the test run compiled and hence fast enough to hopefully catch corner cases.